### PR TITLE
Node consistency 21 30

### PR DESCRIPTION
**Pull request by martinstacey**
### Purpose

The following PR addresses nodes' names and descriptions to make them more descriptive for beginners.

**21) NODE: List.MaximumItemByKey**
Output: maximumItem
Output Description: maximum item in list using keyProjector

**22) NODE:  List.Scan**
-CHANGES: 
Output: scanned to list
Inputs: list1 to list0
Input Description: list1 to List #0

**23) NODE:  List.Reduce**
-CHANGES: 
Output: reduced to list
Inputs: list1 to list0
Input Description: list1 to List #0

**24) NODE:  List.Permutations**
-CHANGES: 
Output: perm to permutations
CHECKED:
its working in Automatic lacing

**25) NODE:  List.Create**
-CHANGES: No changes needed

**26) NODE:  File From Path**
-CHANGES: 
Input description: Path to the file as string,

**27) NODE:  Geometry.ToSolidDef** 
-FYI: Node’s description is inside Protogeometry.dll should be changed by Autodesk

**28) NODE:  Curve.Length**
-FYI: Node’s description is inside Protogeometry.dll should be changed by Autodesk

**29) NODE:  Curve.IsPlanar**
-FYI: Node’s description is inside Protogeometry.dll should be changed by Autodesk

**30) NODE:  Curve.IsClosed**
-FYI: Node’s description is inside Protogeometry.dll should be changed by Autodesk

image

### Declarations

Check these if you believe they are true

- [x] The codebase is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [x] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [x] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated

### Reviewers
@SHKnudsen
@Amoursol 
@QilongTang 

### FYIs
-	Geometry.ToSolidDef, Curve.Length, Curve.IsPlanar, Curve.IsClosed  : Node is inside Protogeometry.dll should be changed by Autodesk

### CODE REVIEW

**Comment by Amoursol** [comment]
@martinstacey Please provide improvement suggestions for all the nodes that are locked into ProtoGeometry behind ADSK VPN.

**Comment by Amoursol** [comment]
@martinstacey Please provide improvement suggestions for all the nodes that are locked into ProtoGeometry behind ADSK VPN.

**Review by QilongTang — APPROVED** [review]

File: src/DynamoCore/BuiltInAndOperators/BuiltIn.xml [hunk 1]
```diff
@@ -76,6 +76,7 @@
       <summary>Returns the items from the list that generate the maximum values for the function supplied as the key projector</summary>
       <param name="list">list of values</param>
       <param name="keyProjector">function applied to the list items</param>
+      <returns name="maximumItem">maximum item in list using keyProjector</returns>
```
Comment (aparajit-pratap) [inline src/DynamoCore/BuiltInAndOperators/BuiltIn.xml:79]: include type `var[]..[]` in description

File: src/Libraries/CoreNodeModels/HigherOrder/MapCombineLacing.cs [hunk 2]
```diff
@@ -283,9 +283,9 @@ public Reduce()
         {
             InPorts.Add(new PortModel(PortType.Input, this, new PortData("reductor", Resources.ReducePortDataReductorToolTip)));
             InPorts.Add(new PortModel(PortType.Input, this, new PortData("seed", Resources.ReducePortDataSeedToolTip)));
-            InPorts.Add(new PortModel(PortType.Input, this, new PortData("list1", Resources.PortDataListToolTip + " #1")));
+            InPorts.Add(new PortModel(PortType.Input, this, new PortData("list0", Resources.PortDataListToolTip + " #0")));
 
-            OutPorts.Add(new PortModel(PortType.Output, this, new PortData("reduced", Resources.ReducePortDataResultToolTip)));
+            OutPorts.Add(new PortModel(PortType.Output, this, new PortData("list", Resources.ReducePortDataResultToolTip)));
```
Comment (aparajit-pratap) [inline src/Libraries/CoreNodeModels/HigherOrder/MapCombineLacing.cs:288]: I think we should have types specified for these node ports as well for consistency with other nodes.

File: src/Libraries/CoreNodes/List.cs [hunk 3]
```diff
@@ -1217,7 +1217,7 @@ public static IList Shuffle(IList list)
         /// </summary>
         /// <param name="list">List to permute.</param>
         /// <param name="length">Length of each permutation.</param>
-        /// <returns name="perm">Permutations of the list of the given length.</returns>
+        /// <returns name="permutations">Permutations of the list of the given length.</returns>
```
Comment (aparajit-pratap) [inline src/Libraries/CoreNodes/List.cs:1220]: Include return type in description?

File: src/DynamoCore/BuiltInAndOperators/BuiltIn.xml (hunk 1 above)
Comment (martinstacey) [reply src/DynamoCore/BuiltInAndOperators/BuiltIn.xml]: addressed in f7c631c

File: src/Libraries/CoreNodes/List.cs (hunk 3 above)
Comment (martinstacey) [reply src/Libraries/CoreNodes/List.cs]: addressed in 5a23f24

File: src/Libraries/CoreNodeModels/HigherOrder/MapCombineLacing.cs (hunk 2 above)
Comment (martinstacey) [reply src/Libraries/CoreNodeModels/HigherOrder/MapCombineLacing.cs:288]: Addressed in cbff7ad
Output type specified in node description.

**Comment by martinstacey** [comment]
> @martinstacey Please provide improvement suggestions for all the nodes that are locked into ProtoGeometry behind ADSK VPN.

@Amoursol 

Suggestions for nodes on Protogeometry:

**Curve.Length**
Input description curve: a curve
Output description double: total arc length of curve

**Curve.IsPlanar**
Input description curve: a curve
Output description bool: is curve planar

**Curve.IsClosed**
Input description curve: a curve
Output description bool: is curve on closed loop

**Comment by Amoursol** [comment]
F

> > @martinstacey Please provide improvement suggestions for all the nodes that are locked into ProtoGeometry behind ADSK VPN.
> 
> @Amoursol
> 
> Suggestions for nodes on Protogeometry:
> 
> **Curve.Length**
> Input description curve: a curve
> Output description double: total arc length of curve
> 
> **Curve.IsPlanar**
> Input description curve: a curve
> Output description bool: is curve planar
> 
> **Curve.IsClosed**
> Input description curve: a curve
> Output description bool: is curve on closed loop

JIRA task for LibG improvements added here: https://jira.autodesk.com/browse/DYN-3145

**Comment by Amoursol** [comment]
@martinstacey - please also change the input port on **List.MaximumItemByKey**: 'keyProjector' --> 'keyFunction'

**Comment by martinstacey** [comment]
> @martinstacey - please also change the input port on **List.MaximumItemByKey**: 'keyProjector' --> 'keyFunction'

addressed in 1d5ec31

**Review by Amoursol — APPROVED** [review]

**Comment by Amoursol** [comment]
@martinstacey Please change **List.Scan** and **List.Reduce** inputs of `reductor` --> `reductorFunction`

**Comment by aparajit-pratap** [comment]
@martinstacey this PR causes the following test to fail:
`DynamoCoreWpfTests.JSONSerializationTests.NewCustomNodeSaveAndLoadPt2`
Please let us know if you're not able to fix it.

**Comment by aparajit-pratap** [comment]
> @martinstacey Please change **List.Scan** and **List.Reduce** inputs of `reductor` --> `reductorFunction`

How about `reduceFunction`?

**Comment by Amoursol** [comment]
> > @martinstacey Please change **List.Scan** and **List.Reduce** inputs of `reductor` --> `reductorFunction`
> 
> How about `reduceFunction`?

This is indeed better! Please amend @martinstacey :)

**Comment by martinstacey** [comment]
> > > @martinstacey Please change **List.Scan** and **List.Reduce** inputs of `reductor` --> `reductorFunction`
> > 
> > 
> > How about `reduceFunction`?
> 
> This is indeed better! Please amend @martinstacey :)

Addressed in c3e0f5c

**Comment by martinstacey** [comment]
> @martinstacey this PR causes the following test to fail:
> `DynamoCoreWpfTests.JSONSerializationTests.NewCustomNodeSaveAndLoadPt2`
> Please let us know if you're not able to fix it.

@aparajit-pratap With my last PR: c3e0f5c checks on GitHub were successful, nevertheless I got this message while doing the test on my computer, any ideas on how to fix it?
image

**Comment by mjkkirschner** [comment]
I believe PT.1 of that test creates a customNode, try deleting it from your definitions folder (or where ever the test created it)

**Comment by martinstacey** [comment]
> I believe PT.1 of that test creates a customNode, try deleting it from your definitions folder (or where ever the test created it)

Ok I just runned successfully first  part 1 and then part 2 of the tests, this PR should be ready for merge

image

**Review by aparajit-pratap — APPROVED** [review]

**Pull request merged**